security/pkg/server/ca/server.go

```diff
@@ -101,6 +101,25 @@ func (s *Server) CreateCertificate(ctx context.Context, request *pb.IstioCertifi
 	return response, nil
 }
 
+// RootCertExpiration returns the remaining duration that root cert is still valid, if already expires, returns error.
+// ========================================
+// TODO: anyone who plan to support ROOT ROTATION FEATURE or MULTI ROOT FEATURE
+// ! PLEASE ENSURE THE ROOT CERT CAN BE ACCESSED SAFELY CONCURRENTLY !
+// ========================================
+func (s *Server) rootCertExpirationSeconds() float64 {
+	rb := s.ca.GetCAKeyCertBundle().GetRootCertPem()
+	cert, err := util.ParsePemEncodedCertificate(rb)
+	if err != nil {
+		log.Errorf("Failed to parse the root cert: %v", err)
+		return -1
+	}
+	end := cert.NotAfter
+	if end.Before(time.Now()) {
+		log.Errorf("Expired Citadel Root found, x509.NotAfter %v, please checkout <TODO-LINK-TO-USER-GUIDE>", end)
```

Comment:
instead of error log, how about just panic to exit directly ?  so that ppl would get notified when see citadel keep crashing with expired root..

Reply:
i thought about it and didn't do it because then the function is not unit testable; also the idea is the metrics going small is to warn people;  for now panic and restart does not help with the situation and people need manual intervention anyway; also when the citadel comes up after panic, it will crash immediately, triggered by a Prometheus scrape call.